### PR TITLE
Implement better CSS parsing to fix relative links.

### PR DESCRIPTION
**Pull request by meenie**
@webbird: Can I get your thoughts on this?
@TomasVotruba: I'm also pinging you on this because you seem a bit active around Munee at the moment :).

What I'm doing is using a new library called [PHP-CSS-Parser](https://github.com/sabberworm/PHP-CSS-Parser) to parse the CSS and fix the `@import` urls and image urls. I think it's better (and safer) to take this route rather than using RegEx.  I've also taken into consideration if someone has skipped using URL Rewrite and used Munee directly.  For example, you can run css through like `/munee.php?files=/foo/bar.css`.  In the CSS it will change imports from `@import url(baz.css)` to `@import url(/munee.php?files=/foo/baz.css)`.

I didn't want to change how CSS itself functions and pull in the actual content and replace the `@import`s like the previous code that @webbird put in.  I thought it would be okay at first, but had a think on it and I think I would rather do this instead.  If someone wants to actually import the content of other CSS files, they should use LESS or SCSS to do so.

What do you guys think?

Here's what the CSS looks like before:
```css
@import 'test2.css';
@import url("test2.css") print;
@import url("test2.css") projection, tv;
@import 'test2.css';
@import url(/foo/test2.css);
@import url("chrome://communicator/skin");
@import "test2.css" screen, projection;
@import url('test2.css') screen and (orientation:landscape);

body {
    color: pink;
    background-image: url('./test.jpg?resize=w[100]');
}
```

And this is after when you request it via `/foo/test.css`
```css
@import url("/foo/test2.css");
@import url("/foo/test2.css") print;
@import url("/foo/test2.css") projection, tv;
@import url("/foo/test2.css");
@import url("/foo/test2.css");
@import url("chrome://communicator/skin");
@import url("/foo/test2.css") screen, projection;
@import url("/foo/test2.css") screen and (orientation:landscape);
body {
    color: pink;
    background-image: url("/foo/test.jpg?resize=w[100]");
}
```

And if you request it via `/munee.php?files=/foo/test.css`
```css
@import url("/munee.php?files=/foo/test2.css");
@import url("/munee.php?files=/foo/test2.css") print;
@import url("/munee.php?files=/foo/test2.css") projection, tv;
@import url("/munee.php?files=/foo/test2.css");
@import url("/munee.php?files=/foo/test2.css");
@import url("chrome://communicator/skin");
@import url("/munee.php?files=/foo/test2.css") screen, projection;
@import url("/munee.php?files=/foo/test2.css") screen and (orientation:landscape);
body {
    color: pink;
    background-image: url("/munee.php?files=/foo/test.jpg&resize=w[100]");
}
```

### CODE REVIEW

**Comment by TomasVotruba** [comment]
@meenie Sure. I personally use router for links, not css/parsing, so can't help.

File: composer.json [hunk 1]
```diff
@@ -25,6 +25,7 @@
         "imagine/imagine": "0.6.2",
         "coffeescript/coffeescript": "1.3.1",
         "meenie/javascript-packer": "1.1",
-        "tubalmartin/cssmin": "~2.4"
+        "tubalmartin/cssmin": "~2.4",
```
Comment (meenie) [inline composer.json]: Test....

File: config/bootstrap.php [hunk 2]
```diff
@@ -14,6 +14,11 @@
 defined('MUNEE_CACHE') || define('MUNEE_CACHE', MUNEE_FOLDER . DS . 'cache');
 // Define default character encoding
 defined('MUNEE_CHARACTER_ENCODING') || define('MUNEE_CHARACTER_ENCODING', 'UTF-8');
+// Are we using Munee with URL Rewrite (.htaccess file)?
+$requestUri = isset($_SERVER['REQUEST_URI']) ? $_SERVER['REQUEST_URI'] : '';
+defined('MUNEE_USING_URL_REWRITE') || define('MUNEE_USING_URL_REWRITE', strpos($requestUri, 'files=') === false);
```
Comment (gabelimon) [inline config/bootstrap.php]:
```php
strpos($requestUri, 'files=') === false
```
should be replaceable with
```php
!strpos($requestUri, 'files=')
```

Comment (meenie) [inline config/bootstrap.php]: If I do that, and `strpos` returns `0`, I'm in trouble :)

Comment (gabelimon) [inline config/bootstrap.php]: Also this isn't a very php-ish way of doing an if. You can do 
```php
if (!defined('MUNEE_USING_URL_REWRITE')) define('MUNEE_USING_URL_REWRITE', !strpos($requestUri, 'files='));
```

File: src/Munee/Asset/Type.php [hunk 3]
```diff
@@ -301,12 +301,16 @@ protected function checkCache($originalFile, $cacheFile)
      */
     protected function generateCacheFile($file)
     {
-        $requestOptions = serialize($this->request->options);
+        $cacheSalt = serialize(array(
+            $this->request->options,
+            MUNEE_USING_URL_REWRITE,
+            MUNEE_DISPATCHER_FILE
```
Comment (gabelimon) [inline src/Munee/Asset/Type.php]: Do these expect keys?

File: config/bootstrap.php (hunk 2 above)
Comment (meenie) [inline config/bootstrap.php]: I dunno...I kind of like having 3 less characters haha.  It's like this in other parts of the file, so I'll keep it.

File: src/Munee/Asset/Type/Css.php [hunk 4]
```diff
@@ -157,107 +159,105 @@ protected function isScss($file)
...
-                }
+        $cssBlocks = $cssDocument->getAllValues();
 
-                $basePathParts = array_slice($basePathParts, $numOfRecursiveDirs);
-                $basePath = implode('/', array_reverse($basePathParts));
+        $this->fixUrls($cssBlocks, $originalFile);
 
-                if (! empty($basePath) && $basePath[0] != '/') {
-                    $basePath = '/' . $basePath;
+        return $cssDocument->render();
```
Comment (gabelimon) [inline src/Munee/Asset/Type/Css.php]: Does `$cssDocument` actually get the changes applied to `$cssBlocks`?

File: config/bootstrap.php (hunk 2 above)
Comment (gabelimon) [inline config/bootstrap.php]: 1) You're right, my bad
2) If it matches the rest of the firle yes you should match.

File: src/Munee/Asset/Type/Css.php (hunk 4 above)
Comment (meenie) [inline src/Munee/Asset/Type/Css.php]: Yup, PHP automatically passes everything by reference.  So when `$cssBlocks` is passed to `fixUrls` and is iterated and modified, those changes reflect back up to `$cssDocument`.

Comment (gabelimon) [inline src/Munee/Asset/Type/Css.php]: Not always no. In function calls and foreach loops you generall have to do `&` unless the object copy does something smart/funny.
```php
php >$ar = array(1,2,3,4);
php > foreach ($ar as $val) {
php { $val = 1;
php { }
php > print_r($ar);
Array
(
    [0] => 1
    [1] => 2
    [2] => 3
    [3] => 4
)```

Comment (meenie) [inline src/Munee/Asset/Type/Css.php]: It's using a method to assign the value, not assigning it directly.  In any case, this does work fine :).

Comment (gabelimon) [inline src/Munee/Asset/Type/Css.php]: Sounds good :)

**Comment by meenie** [comment]
Merging this since there was no more feedback.  I'll be doing a release now.

**Comment by webbird** [comment]
Sorry for being late, I was offline this week.

" If someone wants to actually import the content of other CSS files, they should use LESS or SCSS to do so."

I don't think so. Using a preprocessor to only avoid imports is kind of crack a nut with a sledgehammer. 

My personal focus was to avoid separate loads to increase performance; that's why I replace the `@import`s with the contents of the imported files. As I am a CMS developer, I have to handle pure CSS files from different sources (templates, modules, CDN, ...), which means a potentially long list of files that are beyond my influence, may or may not be compressed, may or may not contain `@imports`, and so on. So just replacing the imports with munee calls (or require the use of a preprocessor) is not a solution for me. (You may read some articles about available Wordpress solutions for deeper insight if you wish. While our CMS has nothing to do with WP, we face the same challenge.)

However, it is not important for me to have "my" solution in the official Munee release. And, of course, there _is_ an unsolved problem concerning the caching as @meenie mentioned. On the other hand, I can solve that by offering a "cleanup cache" button in my CMS' admin panel if I don't find a better solution. ;)

Just one final question: I think the best way to add "my" solution would be to use a filter. (For example `...?minify=true&imports=true`) Could this be combined with the latest changes made in Munee? If yes - how?

Thank you.

**Comment by meenie** [comment]
@webbird:  Unfortunately, you are not able to register a filter outside of Munee.  I would be willing to add this as part of the CSS Asset Type as you were doing before but put it behind a config setting that is defaulted to false.  Then in your CMS you can switch that option to true in the dispatcher file.  This would be similar to the `lessifyAllCss` and `scssifyAllCss` settings here: http://mun.ee/Usage_Instructions/CSS

But yes, we'll need to get around the caching issue.  I don't actually think it will be that hard to do so either.  Basically, you need to end up with an array that has two keys.  One called `compiled` and the other called `files`.  `compiled` is the end result of pulling in all of the `@import` tags and then `files` is a sub-array of all of the file paths and names as the key and `filemtime()` of the file as a value.  What this allows the cache system to do is when it pulls out the content, it will run through all of those listed files and check the current `filemtime()` with the cached version and see which is newer.  If it's newer, the cache is busted for that file and it's recompiled.  So the array would look something like:
```php
$cache = array(
    'compiled' => '<fully compiled css>',
    'files' => array(
        'file1.css' => '1427130947',
        'file2.css' => '1427130943',
        ...
    )
);
```

**Comment by webbird** [comment]
That solution would do it, thanks. This way, everyone can choose the solution of his choice. 

The array solution sounds good, but where will the array be created, where will it be stored?

**Pull request merged**